### PR TITLE
Recipe App - Group Project

### PR DESCRIPTION
**Pull request by Danie12345**
In this PR we:

- An app that has:
  - A login page.
  - A registration page.
  - A food list (with all CRUD implementation, except for 'update').
  - A recipes list (with all CRUD implementation, except for 'update').
  - A general shopping list view (all missing food for all your recipes and total price).
  - A list of all public recipes from other users with their names and total prices.
- The app contains the following views (plus sign in, sign up, etc.):
  - Foods (inventory of foods of the user): Available only for signed-in users, allows CRUD for owners
  - Recipes (list of all recipes): Available for the signed-in owner of the recipes, allows CRUD for owners
  - General Shopping List (list of ingredients left to buy for all the user's recipes): Available only for signed-in users
  - Public Recipes (list of all public recipes): Available for any user
  - Edit/Modify a recipe ingredient: Available only to the recipe owner
- Steps:
  - Login page and registration page:
    - Built with Devise.
  - Food list:
    - Displays a list of food added by the logged-in user as in the wireframe (for teams with only 2 people - display also quantity of a given food).
    - Leads to a form that allows users to add new food.
  - Recipes list:
    - Displays a list of recipes created by the logged-in user as in the wireframe.
    - Leads to recipe details.
    - If the user is the owner of the recipe, should allow the user to delete it.
  - Public recipe list:
    - Displays a list of all public recipes ordered by the newest as in the wireframe.
    - Leads to recipe details.
  - Recipe details:
    - If the recipe is public or the user is the owner of the recipe, it displays the recipe details as in the wireframe.
    - If the user is the owner of the recipe, leads to the form that allows the user to add new food.
  - General shopping list view (**only for a group that has less than 3 members**):
    - Shows the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    - Counts the total food items and total price of the missing food.
  - Made sure there are no N+1 queries happening.
  - Created a navigation menu that allows users to open all of the pages you created.

### CODE REVIEW

File: spec/models/food_spec.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Food, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (Whoistolu) [inline spec/models/food_spec.rb]: - [x] [REQUIRED] Hi team, kindly ensure you write proper units and integration tests for this project as stated in the project requirement.

File: app/controllers/recipes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  def show
+    @user = current_user
+    @recipe = Recipe.find(params[:id])
+    @foods = @recipe.foods.includes(:recipe_foods).order(created_at: :desc)
+    @data = {}
+    @foods.each do |food|
+      quantity = food.recipe_foods.first.quantity
+      cost = quantity * food.price
+      @data[food.name] = { quantity:, cost:, recipe_food_id: food.recipe_foods.first.id }
```
Comment (Whoistolu) [inline app/controllers/recipes_controller.rb:15]: - [x] [REQUIRED] Kindly ensure that this error is fixed as I can not properly navigate through the application.
Screenshot from 2022-09-02 23-32-36

**Review by Whoistolu — CHANGES_REQUESTED** [review]
Hi Team
@Danie12345 and @devMahmoud 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ]  _Check the inline comments under the review for sections to improve._

### Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

### N/A

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/migrate/20220830164121_create_join_table_recipes_foods.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  def change
+    create_table :recipe_foods do |t|
+      t.decimal :quantity
+      t.references :recipe, null: false, foreign_key: true
+      t.references :food, null: false, foreign_key: true
+
+      t.timestamps
+    end
+  end
+end
```
Comment (phelian23) [inline db/migrate/20220830164121_create_join_table_recipes_foods.rb]: - [x] Kindly fix the naming of this migration file. While trying to migrate the the tables, I got this error:
Capture

**Review by phelian23 — CHANGES_REQUESTED** [review]
Hi @Team,

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Great work installing the required gems
Descriptive PR

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Sign up" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (tresorsawasawa) [inline app/views/devise/registrations/new.html.erb:34]: - [ ] As mentioned by the previous reviewers, kindly make sure the user is able to navigate to your application. Right now when a user tries to **Sign Up** the **Confirmation** action doesn't work so impossible to navigate to the whole application without being authenticated.

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUEST CHANGES :recycle::recycle: 

Hi @Team
Great work on making the changes requested by the previous reviewer  👏👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost-there

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Your design looks good :100: :heavy_check_mark: 
- Good PR description :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb (hunk 4 above)
Comment (devMahmoud) [reply app/views/devise/registrations/new.html.erb:34]: Hi @tresorsawasawa 
You can find the verification link in the console.

Comment (devMahmoud) [reply app/views/devise/registrations/new.html.erb:34]: https://user-images.githubusercontent.com/18054628/188271770-9b29d5e3-20b1-45f7-87fc-ed16e7e9f1c6.mp4

Comment (tresorsawasawa) [reply app/views/devise/registrations/new.html.erb:34]: Hi @devMahmoud 

Yes, 
Imagine you deploy your project online, how will the user be able to **Sign Up**?

Suit on slack

Comment (devMahmoud) [reply app/views/devise/registrations/new.html.erb:34]: Hi @tresorsawasawa 
For development, It's ok to have the verification link in the console.
Also, the requirements didn't specify how to verify the user account.

Comment (tresorsawasawa) [reply app/views/devise/registrations/new.html.erb:34]: Hi @devMahmoud 

You are true. I am really sorry for that. You can make another request review

Comment (devMahmoud) [reply app/views/devise/registrations/new.html.erb:34]: Ok, I will ask for another review
thanks

**Review by elmar8287 — APPROVED** [review]
Hi, 

# APPROVED  👍 🥇 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

Congratulations! 🎉 

### Highlights

- All suggestions are implemented 👍 

### Optional suggestions

- N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
### Please, remember to tag me (@elmar8287) in your question so I can receive the notification 👍 
giphy

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**